File: inventory.cpp
#include <vector>
#include <list>


struct Size2i
{
    int x = 0;
    int y = 0;
};

class Item
{
public:
    Size2i Size;
    int Id = 0;
};

class Inventory
{
protected:
    std::vector<Item*>   Items;
    std::vector<int>    BackpackContents;
    Size2i              BackpackSize;

    size_t GetBackpackIndex(int x, int y)
    {
        return y * BackpackSize.x + x;
    }

public:
    Inventory(const Size2i& backpackSize)
        : BackpackSize(backpackSize)
    {
        BackpackContents.resize(BackpackSize.x * BackpackSize.y);

        for (auto& i : BackpackContents)
            i = -1;
    }

    bool ItemCanFit(Size2i position, Item* item)
    {
        for (int y = position.y; y < position.y + item->Size.y; y++)
        {
            if (y >= BackpackSize.y)
                return false;

            for (int x = position.x; x < position.x + item->Size.x; x++)
            {
                if (x >= BackpackSize.x)
                    return false;

                size_t index = GetBackpackIndex(x, y);
                if (BackpackContents[index] != -1)
                    return false;
            }
        }

        return true;
    }

    bool InsertItem(Size2i position, Item* item)
    {
        if (!ItemCanFit(position, item))
            return false;

        for (int y = position.y; y < position.y + item->Size.y; y++)
        {
            for (int x = position.x; x < position.x + item->Size.x; x++)
            {
                size_t index = GetBackpackIndex(x, y);
                BackpackContents[index] = item->Id;
            }
        }
        Items.push_back(item);

        return true;
    }

    Item* RemoveItem(int id)
    {
        Item* item = nullptr;

        for (std::vector<Item*>::iterator itr = Items.begin(); itr != Items.end(); itr++)
        {
            if ((*itr)->Id == id)
            {
                item = *itr;
                Items.erase(itr);
                break;
            }
        }

        if (!item)
            return nullptr;

        for (auto& slot : BackpackContents)
        {
            if (slot == id)
                slot = -1;
        }

        return item;
    }

    Size2i FindAvailableSlot(Item* item)
    {
        for (int y = 0; y < BackpackSize.y; y++)
        {
            for (int x = 0; x < BackpackSize.x; x++)
            {
                Size2i pos = { x,y };
                if (ItemCanFit(pos, item))
                {
                    return pos;
                }
            }
        }

        return Size2i{ -1,-1 };
    }
};

File: inventory.js
/*

Item:
  id
  size:
    x
    y
  pos:
    x
    y
*/

export default class Inventory {
  constructor (width, height, items = []) {
    this.width = width
    this.height = height

    // this will track the list of rects
    this.items = items
    for (const item of items) {
      this.addItem(item)
    }

    // this will track the full grid
    // not really used, but helpful if you want to draw the gird, in some situations (X/O for example)
    this.grid = [...new Array(width)].map(() => [...new Array(height)].fill(0))
  }

  // try to add an item, if available, return new rect or false
  addItem (item) {
    const pos = this.findAvailablePosition(item)

    if (pos) {
      item.pos = pos
      this.items.push(item)

      // update grid, using first char of id
      for (let x = 0; x < item.size.x; x++) {
        for (let y = 0; y < item.size.y; y++) {
          this.grid[y + pos.y][x + pos.x] = item.id[0]
        }
      }
      return item
    }

    // could not add
    return false
  }

  removeItem (id) {
    const itemIndex = this.items.findIndex(i => i.id === id)

    // not found
    if (itemIndex === -1) {
      return false
    }

    // update the grid
    const item = this.items[itemIndex]
    for (let x = 0; x < item.size.x; x++) {
      for (let y = 0; y < item.size.y; y++) {
        this.grid[y + item.pos.y][x + item.pos.x] = 0
      }
    }

    // remove item from this.items
    this.items.splice(itemIndex, 1)

    return true
  }

  // could the item fit in the grid position?
  itemCanFit (position, item) {
    for (let y = position.y; y < position.y + item.size.y; y++) {
      if (y >= this.height) {
        return false
      }

      for (let x = position.x; x < position.x + item.size.x; x++) {
        if (x >= this.width) {
          return false
        }

        if (this.grid[y][x] !== 0) {
          return false
        }
      }
    }

    return true
  }

  // find a x/y position for item or false
  findAvailablePosition (item) {
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const pos = { x, y }
        if (this.itemCanFit(pos, item)) {
          return pos
        }
      }
    }

    // no space
    return false
  }

  // this is just a representation of grid, not really needed
  getText () {
    return this.grid.map(l => l.map(i => i || '□').join(' ')).join('\n')
  }
}
